lib/terminal/server.go

```diff
@@ -0,0 +1,271 @@
+// limitations under the License.
+
+package terminal
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+
+	"github.com/gravitational/teleport/lib/terminal/terminalv1"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	terminalpb "github.com/gravitational/teleport/protogen/teleport/terminal/v1"
+	log "github.com/sirupsen/logrus"
+)
+
+// ServerOpts contains configuration options for a Teleport Terminal service.
+// Note that some options may be passed in JSON format via ConfigInput as long
+// as ReadFromInput is set to true. If that is the case, then ConfigInput takes
+// precedence over previously-configured values, even if the values set by it
+// are empty.
+type ServerOpts struct {
+	// Addr is the bind address for the server, either in "scheme://host:port"
+	// format (allowing for "tcp", "unix", etc) or in "host:port" format (defaults
+	// to "tcp").
+	Addr string `json:"addr"`
+
+	// CertFile, KeyFile and ClientCAs are either file paths or the raw PEM
+	// contents for the corresponding certificates or keys.
+	// If CertFile and KeyFile are present, TLS is enabled.
+	// Additionally, if ClientCAs is present, then mTLS is enabled.
+	CertFile  string   `json:"cert_file"`
+	KeyFile   string   `json:"key_file"`
+	ClientCAs []string `json:"client_cas"`
+
+	// ReadFromInput and ConfigInput configure the (optional) JSON override for
+	// ServerOpts.
+	ReadFromInput bool      `json:"-"`
+	ConfigInput   io.Reader `json:"-"`
+
+	// ConfigOutput, if present, is the destination for the JSON-marshaled
+	// RuntimeOpts that is sent before the server is started.
+	ConfigOutput io.Writer `json:"-"`
+
+	// ShutdownSignals is the set of captured signals that cause server shutdown.
+	ShutdownSignals []os.Signal `json:"-"`
+}
+
+// RuntimeOpts contains the set of ServerOpts that may be dynamically assigned
+// during server start (eg, a chosen random port for the server bind address).
+type RuntimeOpts struct {
+	// C is the exit channel for a running server.
+	C <-chan error `json:"-"`
+	// Addr is the server bind address (eg: "tcp://localhost:1234").
+	Addr string `json:"addr"`
+	// DialAddr is a Dial-friendly address.
+	DialAddr string `json:"-"`
+	TLS      bool   `json:"tls,omitempty"`
+	MTLS     bool   `json:"mtls,omitempty"`
+}
+
+// Server is a combination of the underlying grpc.Server and its RuntimeOpts.
+type Server struct {
+	*grpc.Server
+	*RuntimeOpts
+}
+
+// Start creates and starts a Teleport Terminal service.
+// If ReadFromInput is present, then server a JSON-encoded version of opts is
+// read from ConfigInput, overriding any settings present.
+// If ConfigOutput is present, then a JSON-marshaled RuntimeOpts is sent to it
+// before server start.
+func Start(opts ServerOpts) (*Server, error) {
+	if opts.ReadFromInput {
+		if err := json.NewDecoder(opts.ConfigInput).Decode(&opts); err != nil {
```

Comment:
You could check if `opts` is the zero value, and emit an error or at least log a warning that opts is being overwritten with the JSON input.

Reply:
I think that is a valid scenario. It's up to Alexey, but I kind expect the usual invocation to be simply `tshd -stdin` with all the rest coming via stdin - seems simpler to do programmatically than passing flags.